components/ai_chat/core/browser/ai_chat_service.cc

```diff
@@ -0,0 +1,441 @@
+  receivers_.Add(this, remote.InitWithNewPipeAndPassReceiver());
+  return remote;
+}
+
+void AIChatService::Bind(mojo::PendingReceiver<mojom::Service> receiver) {
+  receivers_.Add(this, std::move(receiver));
+}
+
+void AIChatService::Shutdown() {
+  // Disconnect remotes
+  receivers_.ClearWithReason(0, "Shutting down");
+}
+
+ConversationHandler* AIChatService::CreateConversation() {
+  base::Uuid uuid = base::Uuid::GenerateRandomV4();
+  std::string conversation_uuid = uuid.AsLowercaseString();
+  // Create the conversation metadata
+  {
+    mojom::ConversationPtr conversation =
+        mojom::Conversation::New(conversation_uuid, "", false);
+    conversations_.insert_or_assign(conversation_uuid, std::move(conversation));
+  }
+  mojom::Conversation* conversation =
+      conversations_.at(conversation_uuid).get();
+  // Create the ConversationHandler. We don't persist it until it has data.
+  std::unique_ptr<ConversationHandler> conversation_handler =
+      std::make_unique<ConversationHandler>(
+          conversation, this, model_service_, credential_manager_.get(),
+          feedback_api_.get(), url_loader_factory_);
+  conversation_observations_.AddObservation(conversation_handler.get());
+
+  // Own it
+  conversation_handlers_.insert_or_assign(conversation_uuid,
+                                          std::move(conversation_handler));
+
+  DVLOG(1) << "Created conversation " << conversation_uuid << "\nNow have "
+           << conversations_.size() << " conversations and "
+           << conversation_handlers_.size() << " loaded in memory.";
+
+  // TODO(petemill): Is this neccessary? This conversation won't be
+  // considered visible until it has entries.
+  OnConversationListChanged();
+
+  // Return the raw pointer
+  return GetConversation(conversation_uuid);
+}
+
+ConversationHandler* AIChatService::GetConversation(
+    const std::string conversation_uuid) {
+  auto conversation_handler_it = conversation_handlers_.find(conversation_uuid);
+  if (conversation_handler_it == conversation_handlers_.end()) {
+    return nullptr;
+  }
+  return conversation_handler_it->second.get();
+}
+
+ConversationHandler*
+AIChatService::GetOrCreateConversationHandlerForPageContent(
+    int associated_content_id,
+    base::WeakPtr<ConversationHandler::AssociatedContentDelegate>
+        associated_content) {
+  ConversationHandler* conversation = nullptr;
+  auto conversation_uuid_it =
+      content_conversations_.find(associated_content_id);
+  if (conversation_uuid_it != content_conversations_.end()) {
+    auto conversation_uuid = conversation_uuid_it->second;
+    // Load from memory or database, but probably not database as if the
+    // conversation is in the associated content map then it's probably recent
+    // and still in memory.
+    conversation = GetConversation(conversation_uuid);
+  }
+  if (!conversation) {
+    // New conversation needed
+    conversation = CreateConversation();
```

Comment:
should this case called `CreateConversationHandlerForPageContent`? i see the logic is the same